### ISSUE TITLE
"fisher add jethrokuan/fzf" hangs  -- wrong git command and URL

### ISSUE THREAD

**Issue by osamuaoki** [jorgebucaran/fisher#556]
Thank you for interesting tool but first try with practically default install condition failed.  This may turn down many user, I am afraid.

After careful check, I see this bug was introduced initially 4d953fa ("feat: custom hosted packages via ssh+git; fix #464", 2018-10-13) during 3.0.9 cycle.    Please fix this at 3.3.0

Anyway, here is how this fails as installed.

```
osamu@goofy ~🐟 fisher add jethrokuan/fzf # No mod
created new fishfile in ~/.config/fish/fishfile
^C
^C⏎
There are still jobs active:

   PID  Command
 11401  fish -c "
                echo fetching $url >&2
                command mkdir -p $fisher_config/$pkg $fisher_cache/(command dirname $pkg)
                if test ! -z \"$branch\"
                     command git clone $url $fisher_config/$pkg --branch $branch --depth
                     or echo fisher: cannot clone \"$url\" -- is this a valid url\? >&2
                else if command curl $curl_opts -Ss -w \"\" $url 2>&1 | command tar -xzf-
                    command rm -Rf $fisher_cache/$pkg
                    command mv -f $fisher_config/$pkg/* $fisher_cache/$pkg
                    command rm -Rf $fisher_config/$pkg
                    command cp -Rf {$fisher_cache,$fisher_config}/$pkg
                else if test -d \"$fisher_cache/$pkg\"
                    echo fisher: cannot connect to server -- looking in \"$fisher_cache/$
                    command cp -Rf $fisher_cache/$pkg $fisher_config/$pkg/..
                else
                    command rm -Rf $fisher_config/$pkg
                    echo fisher: cannot add \"$pkg\" -- is this a valid package\? >&2
                end
            " >/dev/null &

A second attempt to exit will terminate them.
Use 'disown PID' to remove jobs from the list without terminating them.
```

So I commented out `fish -c` invocation and also added `echo >&2` to pertinent parts.
Then I tried again to identify problem.  See the log:

```
osamu@goofy ~🐟 fisher add jethrokuan/fzf # No subshell with echo to stderr
fetching https://codeload.github.com/jethrokuan/fzf/tar.gz/master 
mkdir -p /home/osamu/.config/fisher/github.com/jethrokuan/fzf /home/osamu/.cache/fisher/github.com/jethrokuan
git clone https://codeload.github.com/jethrokuan/fzf/tar.gz/master /home/osamu/.config/fisher/github.com/jethrokuan/fzf --branch  --depth 1
fisher: cannot clone "https://codeload.github.com/jethrokuan/fzf/tar.gz/master" -- is this a valid url?
done in 0.36s
```
It only created repo copy directory.  No content was downloaded.

I see 2 problems, here.

* URL used
    * URL for tar.gz is `https://codeload.github.com/jethrokuan/fzf/tar.gz/master` (NOW)
    * URL for shallow clone is standard repo URL `https://github.com/jethrokuan/fzf` (CORRECT ONE)
* git syntax: command option must be placed correctly (I see no reason to have branch since it is null string here) --> Reorder and drop

Let me showhow this git should have been invoked and files to be downloaded.
```
osamu@goofy ~🐟 git clone --depth 1 https://github.com/jethrokuan/fzf /home/osamu/.config/fisher/github.com/jethrokuan/fzf 
Cloning into '/home/osamu/.config/fisher/github.com/jethrokuan/fzf'...
remote: Enumerating objects: 17, done.
remote: Counting objects: 100% (17/17), done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 17 (delta 1), reused 5 (delta 0), pack-reused 0
Unpacking objects: 100% (17/17), done.
```
I got files!

For URL problem, AWK script needs to be adjusted for curl or git-shallow-clone.

As for code `if test ! -z \"$branch\"` seen in current code, I think branch string length should not be used for deciding download method to be curl or git-shallow-clone.  For downloading with git shallow clone, a feature to select branch should exist. So what is needed is method to specify curl or git-shallow-clone.

I hope this helps.

Osamu

**Comment by jorgebucaran** [comment]
Hi @osamuaoki! Thank you, for digging into this. This is definitely technical debt. 

If you have a fix for this, I'm happy to review your PR -- otherwise we can go ahead and get rid of this feature. I think it's fair to not support the original use case.

**Comment by osamuaoki** [comment]
Hi

I think use of the latest git shallow clone itself over downloading tarball is good approach if it is implemented properly.  Unfortunately, this seems not the case.  Removing this feature is certainly one easy way to work around this nasty situation

But I didn’t want to introduce another regression either. 

So I want to hear from the party who created this feature what problem they were solving by introducing this code set.  That’s why I identified the initial commit.

Also I haven’t thoroughly checked how background process is used exactly for parallel execution etc.

Any way, this is my first month trying fish.  I have been interactive bash + script in dash person :-)

Fyi: I have made wrapper for midnight commander (mc).

Sent from iPhone

**Comment by osamuaoki** [comment]
Hi,

I tried to understand what is going on ... and let me think out loud... here is my additional memo

I am reading fisher doc and I see `@` is used to denote tag to be downloaded for "add" command.

This choice of syntax only make sense if you are thinking only HTTPS access git repo.  But it seems to complicate processing for git repo URL such as `git@github.com:edc/bass.git` as reported in #450.

For the record, how to download particular tag version with git clone:
Also for shallow clone, URL with @... at the end doesn't work:
```
osamu@goofy ~/tmp🐟 git clone --depth 1  git@github.com:jethrokuan/fzf.git@0.15.1
Cloning into 'fzf.git@0.15.1'...
fatal: remote error: 
  jethrokuan/fzf.git@0.15.1 is not a valid repository name
  Email support@github.com for help
```
To get 0.15.1, we do:
```
osamu@goofy ~/tmp🐟 git clone --depth 1  -b 0.15.1 git@github.com:jethrokuan/fzf.git
Cloning into 'fzf'...
remote: Enumerating objects: 18, done.
remote: Counting objects: 100% (18/18), done.
remote: Compressing objects: 100% (16/16), done.
remote: Total 18 (delta 0), reused 10 (delta 0), pack-reused 0
Receiving objects: 100% (18/18), 4.39 KiB | 4.39 MiB/s, done.
Note: checking out 'd165e1f7cbf068b3508cf7c78719e28a11d27995'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>
```
This can also be done for normal clone without `--depth 1`.

If you are to track upstream, tracking real branch is better than using tag like above.  Then `git pull` should work better (need to be verified).

Of course, curl should work with vendor special `https://codeload.github.com/` URL.  But except for `master`, it doesn't work easily. (At least it looked so for github.)

I think the previous code base which relies on vendor specific tarball downloading API is very fragile infrastructure.  Just using the standard git API is more robust.  Also introducing syntax complication does more harm than needed.

I think better approach is downloading source tree with git-clone and define more robust simple syntax for `fisher add`  (No AWK code, please):

Basics case is:
```
> fisher add GIT_REPO_URL
```
If branch or tag needs to be specified as BRANCH
```
> fisher add -b BRANCH GIT_REPO_URL
```
Without -b, BRANCH, fisher should assumes "master" as BRANCH

If GIT_REPO_URL has simple `username/reponame` without `https://` nor `@` nor `:`, let prepend it with `https://github.com/` for convenience.

If multiple GIT_REPO_URLs are used
```
> fisher add GIT_REPO_URL1  GIT_REPO_URL2
```
or 
```
> fisher add -b BRANCH1 GIT_REPO_URL1  -b BRANCH2 GIT_REPO_URL2
```
If  -b is not used and GIT_REPO_URL is like `~/path/to/repo` or `./path/to/repo` or `/path/to/repo`, this should be considered local file tree.

 (I don't like -b syntax but I can't come up with better one.)

It looks like I need some time to think more ...

**Comment by osamuaoki** [comment]
If `GIT_REPO_URL` has simple `site/url/username/reponame` (more than 1 `/`) without `https://` nor `@` nor `:`, let prepend it with `https://` for convenience, too.

**Comment by jorgebucaran** [comment]
@osamuaoki #464 is technical debt and I'm happy to remove the feature. Send me a PR to get rid of it? :)

**Comment by osamuaoki** [comment]
I checked it works at 3.0.8.  So it was working.  But rebasing to drop problematic code was nightmare due to major changes happened.  So I gave up.

I actually think fixing git shallow clone is better and cleaner solution.  I have a proof-of-concept for github repo access with attached patch.  (some bugs exist in this, I know.)

[test.zip](https://github.com/jorgebucaran/fisher/files/3776100/test.zip)

Maybe next weekend ...
